zerver/webhooks/thinkst/view.py

```diff
@@ -0,0 +1,52 @@
+# Webhooks for external integrations.
+from typing import Any, Dict, Optional
+
+from django.http import HttpRequest, HttpResponse
+
+from zerver.decorator import api_key_only_webhook_view
+from zerver.lib.request import REQ, has_request_variables
+from zerver.lib.response import json_success
+from zerver.lib.webhooks.common import check_send_webhook_message
+from zerver.models import UserProfile
+
+
+@api_key_only_webhook_view('Thinkst')
+@has_request_variables
+def api_thinkst_webhook(request: HttpRequest, user_profile: UserProfile,
+                        message: Dict[str, Any]=REQ(argument_type='body'),
+                        user_specified_topic: Optional[str]=REQ('topic', default=None)) -> HttpResponse:
+
+    body = None
+    topic = None
+
+    # There are two types of requests that Thinkst will send - depending on whether it is
+    # a canary or a canarytoken. Unfortunately, there isn't a great way to differentiate other
+    # than look at the contents.
+    is_canarytoken = 'Timestamp' not in message
+
+    if is_canarytoken:
+        topic = 'canarytoken alert'
+        body = (f"**:alert: Canarytoken has been triggered on {message['time']}!**\n\n"
+                f"> {message['memo']} \n\n"
+                f"[Manage this canarytoken]({message['manage_url']})")
+    else:
+        topic = f"canary alert - {message['CanaryName']}"
+
+        reverse_dns = ''
+        if 'ReverseDNS' in message:
+            reverse_dns = f" (`{message['ReverseDNS']}`)"
+
+        name = ''
+        if user_specified_topic:
+            name = f" `{message['CanaryName']}`"
+
+        body = (f"**:alert: Canary{name} has been triggered!**\n\n"
+                f"On {message['Timestamp']}, `{message['CanaryName']}` was triggered "
+                f"from `{message['SourceIP']}`{reverse_dns}:\n\n"
+                f"> {message['Intro']}")
```

Comment:
I think that the bodies of the `if` and `else` blocks should be split out into their own functions based on the type of the event. The functions would take the message and return the respective topic and body.

Reply:
Fixed, thanks!